### PR TITLE
add authorization check for datasource api

### PR DESCRIPTION
**Pull request by TrafalgarRicardoLu**
### What problem does this PR solve? 
This pr related to [issue2088](https://github.com/pingcap/tispark/issues/2088)
Write could success without corresponding privilege
And we also find read don't have authorization check, thought we don't recommend to use it.

### What is changed and how it works?
Do authorization check before writing and reading data by datasource api. 

#### Write
There are two statement 'replace' and 'insert on duplicate key' in TiDB which is quite similar to write in TiSpark when replace is enabled.

But  they are different in TiDB at following situation:

Create table test and insert two records.
```
CREATE table test2 ( 
a int, 
b int,
c int,
d int,
unique key(a), unique key(b));

INSERT INTO test values(1,1,1,1);
INSERT INTO test values(2,2,2,2);

 a | b | c | d
---+---+---+---
 1 | 1 | 1 | 1
 2 | 2 | 2 | 2
```

If we use 'insert into ... duplicate key update ...' statement.
```
INSERT INTO test values(1,2,3,4) on DUPLICATE KEY UPDATE a=1,b=2,c=3,d=4;

error: mysql: 1062: Duplicate entry '2' for key 'b'

it won't be executed.
```
if we use 'replace into ...' statement.
```
REPLACE INTO test values(1,2,3,4);

 a | b | c | d
---+---+---+---
 1 | 2 | 3 | 4

The origin two records will be deleted and insert a new record.
```

TiSpark have similar result with REPLACE statement. So we chose to follow the privilege requirement of REPLACE in TiDB.

In short, this pr add authorization check in write as follows:
| Replace | Privilege |
| -------- | -------- |
| Enable | INSERT and DELTE |
| Unable | INSERT |

#### Read
Read requires SELECT privilege. Since it will generate corresponding object in following step, we will do authorization check for this object. 

We find SELECT and DELETE statements also generate same object, but they need SELECT privilege too, so it won't affect   correctness.

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has exported variable/fields change

### CODE REVIEW

**Comment by xuanyu66** [comment]
The auth check should be optional.

File: core/src/main/scala/com/pingcap/tispark/write/TiBatchWrite.scala [hunk 1]
```diff
@@ -83,6 +84,7 @@ class TiBatchWrite(
   @transient private var startMS: Long = _
   private var startTs: Long = _
   private var twoPhaseCommitHepler: TwoPhaseCommitHepler = _
+  @transient private var tiAuthorization: Option[TiAuthorization] = _
```
Comment (xuanyu66) [inline core/src/main/scala/com/pingcap/tispark/write/TiBatchWrite.scala:87]: Since there is  no need to update tiAuthorization, maybe val is better

File: core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala [hunk 2]
```diff
@@ -357,6 +357,15 @@ class TiBatchWriteTable(
     }
   }
 
+  def checkAuthorization(tiAuthorization: Option[TiAuthorization], options: TiDBOptions): Unit = {
+    if (options.replace) {
```
Comment (shiyuhang0) [inline core/src/main/scala/com/pingcap/tispark/write/TiBatchWriteTable.scala:361]: Can you check if the replace has the same semantics as MySQL replace.
mysql replace need insert and delete pri https://dev.mysql.com/doc/refman/8.0/en/replace.html

File: docs/datasource_api_userguide.md [hunk 3]
```diff
@@ -31,11 +31,12 @@ In addition, if `replace` is true, data to be inserted is duplicated before the
 
     - if the primary key or unique index exists in the table, data is updated.
     - if no same primary key or unique index exists, data is inserted.
-
+    - if enable authorization check, users need insert and delete privilege to write.
```
[the remark this answers is not in the document]
Comment (TrafalgarRicardoLu) [reply docs/datasource_api_userguide.md:34]: According to [mysql replace](https://dev.mysql.com/doc/refman/8.0/en/replace.html), select is no needed.

File: docs/authorization_userguide.md [hunk 4]
```diff
@@ -59,8 +59,11 @@ Currently, Supported statements are as follows:
 | SHOW COLUMNS    | ❌ | ❌ |
 | DESCRIBE COLUMN | ❌ | ❌ |
 
+## Compatibility with Data Source API
+
+We only support authorization check in write currently.
```
[the remark this answers is not in the document]
Comment (TrafalgarRicardoLu) [reply docs/authorization_userguide.md:64]: I have tried it. 
df.read doesn't supporting authorization currently.

File: docs/authorization_userguide.md [hunk 5]
```diff
@@ -59,8 +59,18 @@ Currently, Supported statements are as follows:
 | SHOW COLUMNS    | ❌ | ❌ |
 | DESCRIBE COLUMN | ❌ | ❌ |
 
+## Compatibility with Data Source API
+
+We support authorization check in both read and write, but we don't recommend to use read.
```
[the remark this answers is not in the document]
Comment (TrafalgarRicardoLu) [reply docs/authorization_userguide.md:64]: Spark sql is recommended, users could use SELECT statement to read data.

[the remark this answers is not in the document]
Comment (zhangyangyu) [reply docs/authorization_userguide.md:64]: Why not clarify then? Like "but we recommend using spark sql instead of datasource API to read data from TiDB".

**Review by xuanyu66 — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala [hunk 6]
```diff
@@ -45,20 +46,18 @@ case class TiAuthorizationRule(getOrCreateTiContext: SparkSession => TiContext)(
...
-      sa
     case sd @ SetCatalogAndNamespace(catalogManager, catalogName, namespace) =>
       if (namespace.isDefined) {
         namespace.get
           .foreach(TiAuthorization.authorizeForSetDatabase(_, tiAuthorization))
       }
       sd
+    case dr @ DataSourceV2Relation(table, output, catalog, identifier, options) =>
+      if (table.name().nonEmpty) {
+        val tb_name = table.name().split("\\.")
```
[the remark this answers is not in the document]
Comment (TrafalgarRicardoLu) [reply core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala:57]: According to test, identifier sometimes is empty. So I use table here.

[the remark this answers is not in the document]
Comment (zhangyangyu) [reply core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala:57]: Is empty identifier reasonable? @shiyuhang0 @xuanyu66

File: core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala [hunk 7]
```diff
@@ -45,20 +46,18 @@ case class TiAuthorizationRule(getOrCreateTiContext: SparkSession => TiContext)(
...
+        TiAuthorization.authorizeForSelect(tb_name(1), tb_name(0), tiAuthorization)
+      }
+      dr
```
[the remark this answers is not in the document]
Comment (TrafalgarRicardoLu) [reply core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala:60]: I find TiBatchWrite won't generate a unique object here. So i think it's better to do auth check in TiBatchWrite.

File: core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (xuanyu66) [reply core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala]: It's beyond my knowledge. QAQ

**Review by xuanyu66 — APPROVED** [review]

File: core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala [hunk 8]
```diff
@@ -52,10 +53,18 @@ case class TiAuthorizationRule(getOrCreateTiContext: SparkSession => TiContext)(
...
-        val tb_name = table.name().split("\\.")
-        TiAuthorization.authorizeForSelect(tb_name(1), tb_name(0), tiAuthorization)
+    case dr @ DataSourceV2Relation(
+          TiDBTable(_, tableRef, _, _, _),
+          output,
+          catalog,
+          identifier,
+          options) =>
+      if (tableRef.tableName.nonEmpty) {
+        println("check " + tableRef.tableName + " " + tableRef.databaseName)
```
Comment (zhangyangyu) [inline core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala:63]: Remove this line or use log instead plz.

File: docs/authorization_userguide.md [hunk 9]
```diff
@@ -59,8 +59,20 @@ Currently, Supported statements are as follows:
...
+Spark SQL is recommended to read data.
+
+We will do authorization check as follows:
+
+| Operation | Option            | Required Privilege | 
+|-----------|-------------------|-------------------|
+| Write     | Replace is enable | Insert and Delete |   
+| Write     | Replace is unable | Insert            |
```
Comment (shiyuhang0) [inline docs/authorization_userguide.md:72]:
```suggestion
| Write     | Replace is disable | Insert            |
```

File: docs/authorization_userguide.md [hunk 10]
```diff
@@ -59,8 +59,20 @@ Currently, Supported statements are as follows:
 | SHOW COLUMNS    | ❌ | ❌ |
 | DESCRIBE COLUMN | ❌ | ❌ |
 
+## Compatibility with Data Source API
+
+We support authorization check in both read and write, but we don't recommend to use read.
```
Comment (shiyuhang0) [inline docs/authorization_userguide.md:64]:
```suggestion
We support authorization check in both read and write, but we recommend using spark SQL instead of Datasource API to read data from TiDB
```

Comment (shiyuhang0) [reply docs/authorization_userguide.md:64]: This may be clearer

File: core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (shiyuhang0) [reply core/src/main/scala/org/apache/spark/sql/catalyst/analyzer/TiAuthorizationRule.scala]: I think it should not be empty, so I don't know why.
But I'm sure get from TiDBTable is a feasible method
